clang/lib/CIR/Dialect/IR/CIRDialect.cpp

```diff
@@ -3306,6 +3326,12 @@ LogicalResult BinOp::verify() {
     return emitError() << "The nsw/nuw flags are applicable to opcodes: 'add', "
                           "'sub' and 'mul'";
 
+  bool complexOps = getKind() == mlir::cir::BinOpKind::Add ||
+                    getKind() == mlir::cir::BinOpKind::Sub;
+  if (isa<mlir::cir::ComplexType>(getType()) && !complexOps)
+    return emitError()
+           << "cir.binop can only represent 'add' and 'sub' on complex numbers";
```

Comment:
one small improvement could be "numbers, use cir.complex.binop instead", or something like that.